org.hl7.fhir.validation/src/main/java/org/hl7/fhir/validation/instance/type/QuestionnaireValidator.java

```diff
@@ -758,6 +808,536 @@ private void validateQuestionnaireResponseItemQuantity(List<ValidationMessage> e
+    boolean ok = true;
+    Element v = answer.getNamedChild("value");
+    if (v != null && v.fhirType().equals("integer")) {
+      NodeStack vns = ns.push(v, -1, null, null);
+      try {
+        int vdt = Integer.parseInt(v.primitiveValue()); 
+        if (qItem.hasExtension(ToolingExtensions.EXT_MINVALUE)) {
+          int dt = Integer.parseInt(qItem.getExtensionByUrl(ToolingExtensions.EXT_MINVALUE).getValue().primitiveValue());
+          ok = rule(errors, "2024-05-07", IssueType.INVARIANT, vns, vdt >= dt, I18nConstants.QUESTIONNAIRE_QR_ITEM_INTEGER_MIN, vdt, dt) && ok;
+        }
+        if (qItem.hasExtension(ToolingExtensions.EXT_MAXVALUE)) {
+          int dt = Integer.parseInt(qItem.getExtensionByUrl(ToolingExtensions.EXT_MAXVALUE).getValue().primitiveValue());
+          ok = rule(errors, "2024-05-07", IssueType.INVARIANT, vns, vdt <= dt, I18nConstants.QUESTIONNAIRE_QR_ITEM_INTEGER_MAX, vdt, dt) && ok;
+        }
+      } catch (Exception e) {
+        ok = rule(errors, "2024-05-07", IssueType.INVARIANT, vns, false, I18nConstants.QUESTIONNAIRE_QR_ITEM_INTEGER_VALUE_EXCEPTION, e.getMessage()) && ok;
+      }
+      if (qItem.getAnswerOption().size() > 0) {
+        List<IntegerType> list = new ArrayList<IntegerType>();
+        for (QuestionnaireItemAnswerOptionComponent components : qItem.getAnswerOption()) {
+          try {
+            list.add(new IntegerType(components.getValue().primitiveValue()));
+          } catch (FHIRException e) {
+            // If it's the wrong type, just keep going
+          }
+        }
+        if (list.isEmpty()) {
+          ok = rule(errors, "2024-05-07", IssueType.INVARIANT, vns, false, I18nConstants.QUESTIONNAIRE_QR_ITEM_NOOPTIONSINTEGER) && ok;
+        } else {
+          boolean found = false;
+          for (IntegerType item : list) {
+            if (item.primitiveValue().equals(v.primitiveValue())) {
+              found = true;
+              break;
+            }
+          }
+          if (!found) {
+            ok = rule(errors, "2024-05-07", IssueType.INVARIANT, vns, found, I18nConstants.QUESTIONNAIRE_QR_ITEM_NOINTEGER, v.primitiveValue()) && ok;
+          }
+        }
+      } else {
+      }
+    }
+    return ok;
+  }
+
+  private int precision(String v) {
+    if (Utilities.noString(v) || !v.contains(".")) {
+      return 0;
+    } else {
+      return v.length() - v.lastIndexOf(".");
+    }
+  }
+
+  private boolean validateQuestionnaireResponseItemReference(List<ValidationMessage> errors, Element answer, NodeStack ns, QuestionnaireItemComponent qItem) {
+    boolean ok = true;
+    Element v = answer.getNamedChild("value");
+    if (v != null && v.fhirType().equals("Reference")) {
+      NodeStack vns = ns.push(v, -1, null, null);
+      String ref = v.getNamedChildValue("reference");
+      if (ref != null && settings.isAssumeValidRestReferences()) {
```

Comment:
Why do we use isAssumeValidRestReferences here? This looks like it's checking against a list of allowed reference types. Shouldn't it check those regardless of whether or not we're assuming valid references?

Reply:
see https://confluence.hl7.org/spaces/FHIR/pages/35718580/Using+the+FHIR+Validator#UsingtheFHIRValidator-assumeValidRestReferences